docs/design/pvc-template.md

```diff
@@ -0,0 +1,246 @@
+    
+While block device is the best volume choice as the scratch space of training job, and k8s has native support for auto
+block device provisioning through [Persistent Volume Claim](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#lifecycle-of-a-volume-and-claim),
+it would be a easy leverage for kubeflow to support dynamic scratch volume provisioning for training jobs.
+
+### User Stories
+- As a data scientist / engineer, I want to use arbitrary disk type and size for my different training jobs
+- As a data scientist / engineer, I want the infrastructure of my training jobs get automatically handled and I don't need to worry about provisioning and recycling them
+- As an infrastructure developer, I don't want to have user application to use host disk space with unlimited space as this increases the risk of using up host disk and affect stability
+- As an infrastructure developer, I want to provision infrastructure to my customers only when needed or there will be a waste of money
+
+### Goals
+- Automatically provision / de-provision block device as scratch space for user in an efficient and reliable way
+- Design should be generic enough so both `PyTorchJob` and `TFJob` can take this feature
+- When a pod gets retried (recreated), it should get it's volume back so it has the option to continue with its work
+- User don't have to define volume for every single replica of worker
+- Volume creation throttling
+    - This has to be added for volume provisioning feature to be production ready since most cloud provider throttles
+      API calls per account - if there is a burst of volume creation and cloud provider started to throttle calls, all
+      clusters under the entire cloud account would get affected, and massive back-offs are hard to handle given the
+      error might propagate through upstream services
+      
+### Non-Goals
+- Volume pooling for reducing Create/Delete calls as we can re-use volume for different jobs
+- Smart throttling such that one training job will not starve other ones due to volume throttling, similar for different
+  worker types within same training job
+
+
+## Proposal
+
+### Implementation Details
+
+I suggest that we start to support dynamic volume provisioning from `v1beta2`, and change will not be back ported to older versions.
+Currently, both tf-operator and pytorch-operator need to be changed.
+
+#### API Change
+Similar to stateful set, we will add a field in replica spec to define PVC template:
+```go
+// ReplicaSpec is a description of the replica
+type ReplicaSpec struct {
+	// ...
+	
+	// VolumeClaimTemplates specifies a list of volume claim templates, which defines volumes specified in template.
+	VolumeClaimTemplates []corev1.PersistentVolumeClaim `json:"volumeClaimTemplates,omitempty"`
+}
+```
+
+Since both `TFJob` and `PyTorchJob` will use `common.ReplicaSpec` since `v1beta2`, such change can be easily adopted by
```

Comment:
Agree that since we are modifying `common.ReplicaSpec`, this feature should be generic enough so any distributed training job can use it.

Regarding versions, what is your opinion on back-porting volume support to beta versions?